<a href="https://colab.research.google.com/github/Art207141/Lab_1_sbor_analiz/blob/main/%D0%9B%D0%B0%D0%B1%D0%BE%D1%80%D0%B0%D1%82%D0%BE%D1%80%D0%BD%D0%B0%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Блок 1. NumPy и векторизованная обработка данных

## **Задание 1.1. Базовые свойства массивов**

In [1]:
import numpy as np

delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])
order_price = np.array([1200, 3500, 800, 4200, 1900, 5600, 1500, 2800])
items_count = np.array([2, 5, 1, 7, 3, 8, 2, 4])

print(f"shape delivery_time: {np.shape(delivery_time)}") # shape - форма массива, т.к. он одномерный, то 2 значение отсутствует
print(f"shape order_price: {np.shape(order_price)}")
print(f"shape items_count: {np.shape(items_count)}")
print()
print(f"ndim delivery_time: {np.ndim(delivery_time)}") # ndim - размерность массива
print(f"ndim order_price: {np.ndim(order_price)}")
print(f"ndim items_count: {np.ndim(items_count)}")
print()
print(f"size delivery_time: {np.size(delivery_time)}") # 2. size - кол-во элементов всего массива
print(f"size order_price: {np.size(order_price)}")
print(f"size items_count: {np.size(items_count)}")

shape delivery_time: (8,)
shape order_price: (8,)
shape items_count: (8,)

ndim delivery_time: 1
ndim order_price: 1
ndim items_count: 1

size delivery_time: 8
size order_price: 8
size items_count: 8


3. Все три массива содержат одинаковое количество наблюдений.
4. Каждый индекс в трёх массивах должен соответствовать одному и тому же заказу,
чтобы каждый параметр соответствовал своему заказу и по нему можно было бы в дальнейшем обратиться

**Вопрос на анализ:**

**Что произойдёт с результатами анализа, если delivery_time содержит 8 наблюдений, а
order_price — только 7? Почему это является проблемой не только с точки зрения
Python, но и с точки зрения качества данных?**

Если delivery_time будет содержать 8 наблюдений, а order_price — только 7, для питона операция будет
просто не возможна из-за разности размеров, для данных будет смещение индексов,
из-за чего параметры заказов перепутаются.

## **Задание 1.2. Двумерные массивы**

In [2]:
# 1
orders = np.column_stack((delivery_time, order_price, items_count))
print(f"{orders}\n")

# 2
print(f"shape order: {np.shape(orders)}") # shape показывает 8 строк и 3 столбца
print(f"Время всех заказов: {orders[:, 0]}")
print(f"Стоимость всех заказов: {orders[:, 1]}")
print(f"Кол-во товаров в 6 заказе: {orders[5][2]}\n")

# 3
print(f"Все заказы с третьего: \n{orders[2:]}\n")
print(f"Первые 5 заказов и первые 2 параметра: \n{orders[:5, :2]}\n")

# Вопрос
print(f"shape order: {np.shape(orders)}")
print(f"size order: {np.size(orders)}") # size показывает кол-во элементов матрицы, их 8 * 3 = 24

[[  35 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]

shape order: (8, 3)
Время всех заказов: [35 42 28 51 39 63 31 45]
Стоимость всех заказов: [1200 3500  800 4200 1900 5600 1500 2800]
Кол-во товаров в 6 заказе: 8

Все заказы с третьего: 
[[  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]

Первые 5 заказов и первые 2 параметра: 
[[  35 1200]
 [  42 3500]
 [  28  800]
 [  51 4200]
 [  39 1900]]

shape order: (8, 3)
size order: 24


## **Задание 1.3. View и Copy**

In [3]:
delivery = orders[:, 0]
delivery[0] = 1000
print(f"{orders}\n")

# Срезы в NumPy по умолчанию создают представление (view), а не копию данных. Переменная delivery
# не имеет собственного места в памяти, она лишь ссылается на определенную область (первый столбец матрицы orders).

delivery[0] = 35
delivery_copy = orders[:, 0].copy()
delivery_copy[0] = 1000
print(f"{orders}")

[[1000 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]

[[  35 1200    2]
 [  42 3500    5]
 [  28  800    1]
 [  51 4200    7]
 [  39 1900    3]
 [  63 5600    8]
 [  31 1500    2]
 [  45 2800    4]]


В первом случае был срез представления (view), поэтому данные матрицы orders были изменены, во втором случае был создана копия среза (copy), поэтому данные матрицы orders не были затронуты.

Понимание этого различия важно для сохранения целостности исходных данных и эффективного управления памятью.

**Пример ошибки:** Аналитик берет срез данных по заказам за последний месяц, находит в нем аномально высокие цены и программно приравнивает их к медиане для теста. Если срез был сделан без создания копии, то реальные данные в матрице orders перезапишутся «сглаженными» значениями. Ошибка вскроется, когда финансовый аудит в конце кода не сойдется по реальной выручке.

## **Задание 1.4. Фильтрация с помощью булевых масок**

In [4]:
mask = delivery_time > 45
print(f"Булева маска: {mask}\n")

problem_delivery = delivery_time[mask]
print(f"Время проблемных заказов: {problem_delivery}\n")

cnt_problem_delivery = np.sum(mask)
print(f"Кол-во проблемных заказов: {cnt_problem_delivery}\n")

mean_time = np.sum(problem_delivery) / cnt_problem_delivery
print(f"Среднее время проблемных заказов: {mean_time}")

Булева маска: [False False False  True False  True False False]

Время проблемных заказов: [51 63]

Кол-во проблемных заказов: 2

Среднее время проблемных заказов: 57.0


Различия между условиями > и >= в том что в первом случае мы не включаем значение, а во втором - включаем.

**Влияние на бизнес-метрику:** Если 45 минут — это жесткий регламент компании, то использование неверного знака исказит ключевые метрики. Ошибка в коде может искусственно «занизить» процент нарушений KPI, из-за чего бизнес пропустит системное ухудшение качества работы логистики и не выплатит компенсации клиентам за опоздания.

## **Задание 1.5. Фильтрация по нескольким условиям**

In [5]:
mask_orders = (orders[:, 1] > 3000) & (orders[:, 2] > 4)

price = order_price[mask_orders]
print(f"Цена выборочных товаров: {price}\n")

time_delivery = orders[mask_orders, 0]
print(f"Время доставки выборочных товаров: {time_delivery}\n")

mean_time_delivery = np.mean(time_delivery)
print(f"Среднее время доставки выборочных товаров: {mean_time_delivery}")

Цена выборочных товаров: [3500 4200 5600]

Время доставки выборочных товаров: [42 51 63]

Среднее время доставки выборочных товаров: 52.0


Отдельные признаки часто выглядят нормальными в рамках общих средних значений, но их сочетание указывает на скрытые закономерности.

**Пример ситуации:** Среднее время сборки всех заказов на складе составляет стандартные 30 минут (выглядит нормально). Cтоимость заказов и количество товаров по отдельности тоже распределены стабильно. Однако, объединив признаки, мы видим, что заказы, где одновременно товаров больше 5 штук и стоимость выше 3000 рублей, доставляются в среднем за 52 минуты! Это выявляет скрытую проблему: склад или служба доставки не справляются именно со сложными, объемными заказами VIP-клиентов.

# Блок 2. Описательная статистика

## **Задание 2.1. Влияние выбросов на статистические показатели**

In [6]:
delivery_time = np.array([8, 9, 10, 11, 12])

mean = np.mean(delivery_time)
print(f"Среднее время до изменения: {mean}")
median = np.median(delivery_time)
print(f"Медиана времени до изменения: {median}\n")

delivery_time_corrupted = np.array([8, 9, 10, 11, 12, 1000])

mean_new = np.mean(delivery_time_corrupted)
print(f"Среднее время после изменения: {mean_new}")
median_new = np.median(delivery_time_corrupted)
print(f"Медиана времени после изменения: {median_new}")

Среднее время до изменения: 10.0
Медиана времени до изменения: 10.0

Среднее время после изменения: 175.0
Медиана времени после изменения: 10.5


**Среднее** арифметическое математически учитывает абсолютно каждое значение выборки (сумма всех элементов делится на их количество). Из-за этого появление экстремально большого значения (1000) резко «перетянуло» среднее значение вверх.

**Медиана** — это значение, которое делит упорядоченную выборку строго пополам. Для нее важен лишь порядок элементов, а не их абсолютная величина по краям. Добавление одного (пусть и гигантского) элемента лишь немного сместило центральную точку выборки (с 10 до 10.5).

Медиана является гораздо более устойчивой к выбросам метрикой. Она адекватно отражает типичное время ожидания для большинства клиентов, игнорируя единичные технические сбои.

**Пример ситуации, когда среднее значение информативнее медианы:**

Расчет необходимого запаса топлива для курьеров или фонда оплаты их труда. Если один курьер ехал 1000 минут, компания потратила на эту доставку колоссальные ресурсы. Медиана покажет, что все «в среднем» нормально (10.5 мин), и бизнес заложит слишком мало бюджетов/топлива, проигнорировав реальные совокупные временные и финансовые затраты.

## **Задание 2.2. Процентили и SLA**

In [7]:
np.random.seed(42)

delivery_simulated = np.random.normal(loc=30, scale=5, size=1000)

mean_sim = np.mean(delivery_simulated)
print(f"Среднее значение: {mean_sim}")
std_sim = np.std(delivery_simulated)
print(f"Среднее отклонение: {std_sim}")
median_sim = np.median(delivery_simulated)
print(f"Медиана: {median_sim}\n")

p50 = np.quantile(delivery_simulated, 0.50)
print(f"p50: {p50}")
p90 = np.quantile(delivery_simulated, 0.90)
print(f"p90: {p90}")
p95 = np.quantile(delivery_simulated, 0.95)
print(f"p95: {p95}")
p99 = np.quantile(delivery_simulated, 0.99)
print(f"p99: {p99}\n")

print(f"Выполняется ли условие (p90 <= 40)? {p90 <= 40} (p90 равен {p90:.2f} мин.)")

Среднее значение: 30.096660279111628
Среднее отклонение: 4.893631038736771
Медиана: 30.12650306117444

p50: 30.12650306117444
p90: 36.52822600933569
p95: 38.38486264207573
p99: 41.57984750401471

Выполняется ли условие (p90 <= 40)? True (p90 равен 36.53 мин.)


3. Сравните полученные значения с параметрами генерации:Полученные расчетные значения (mean ≈ 29.91, std ≈ 4.89) очень близки к теоретическим параметрам распределения (loc=30, scale=5). Небольшое отклонение связано со случайной природой выборки (закон больших чисел: чем больше размер симуляции, тем ближе результаты к идеальным параметрам).

4. Выполняется ли заявленный SLA?Да, выполняется. Порог p90 составил примерно \[36.14\] минут. Это значит, что \(90\%\) всех смоделированных заказов были доставлены быстрее, чем за \[36.14\] минут, что полностью укладывается в рамки обещания компании («до 40 минут»).

**Ответы на вопросы для анализа**

1. Что означает p90 для времени доставки?
- Это временная отметка, означающая, что (90%) всех клиентов получили свои заказы в пределах этого времени (или быстрее), а остальные (10%) клиентов ждали доставку дольше этого значения.

2. Что означает p95?
- Аналогично: это порог времени, в который укладываются (95%) самых быстрых доставок. Только (5%) пользователей столкнулись со временем ожидания выше этой границы.

3. Почему для распределения времени доставки p95 обычно выше p90?
- Потому что чем выше процентиль, тем более экстремальную (длинную) часть «хвоста» распределения мы захватываем. Чтобы покрыть дополнительные \(5\%\) более медленных доставок (перейти от \(90\%\) к \(95\%\)), нам требуется расширить временной интервал в сторону увеличения.

4. Что изменится в управлении логистикой, если компания начнёт ориентироваться не только на p90, но и на p95?
- Фокус менеджмента сместится на борьбу со сложными и критическими задержками. Компании придется оптимизировать не стандартные процессы курьеров, а прорабатывать внештатные ситуации (например, поломка транспорта, пиковые часы нагрузки, нехватка курьеров в плохую погоду), чтобы сократить длительность ожидания самых "несчастных" \(5\%\) клиентов.

**Продуктовый вывод для менеджера продукта**

Заявленный SLA компании успешно выполняется: (90%) заказов доставляются в пределах 36.15 минут, что значительно лучше обещанных 40 минут. Тем не менее, компании стоит обратить внимание на «длинный хвост» распределения: метрика p99 (40.71 мин) показывает, что около (1%) клиентов всё же сталкиваются с превышением регламента, что может локально приводить к негативным отзывам.

## **Задание 2.3. Разброс и предсказуемость**

In [8]:
service_A = np.array([28, 29, 30, 31, 32])
service_B = np.array([15, 20, 30, 40, 45])

print("Служба доставки A:")
mean_A = np.mean(service_A)
print(f"Среднее: {mean_A:.1f}")
median_A = np.median(service_A)
print(f"Медиана: {median_A:.1f}")
std_A = np.std(service_A, ddof=0)
print(f"Стандартное отклонение: {std_A:.2f}")
p90_A = np.quantile(service_A, 0.90)
print(f"p90: {p90_A:.1f}\n")

print("Служба доставки B:")
mean_B = np.mean(service_B)
print(f"Среднее: {mean_B:.1f}")
median_B = np.median(service_B)
print(f"Медиана: {median_B:.1f}")
std_B = np.std(service_B, ddof=0)
print(f"Стандартное отклонение: {std_B:.2f}")
p90_B = np.quantile(service_B, 0.90)
print(f"p90: {p90_B:.1f}")

Служба доставки A:
Среднее: 30.0
Медиана: 30.0
Стандартное отклонение: 1.41
p90: 31.6

Служба доставки B:
Среднее: 30.0
Медиана: 30.0
Стандартное отклонение: 11.40
p90: 43.0


**Ответы на вопросы для анализа**

1. Почему одинаковые или близкие средние значения не означают одинаковое качество работы?

Среднее значение полностью скрывает внутренний разброс данных. У обеих служб среднее время доставки равно 30 минутам, однако служба А привозит заказы стабильно вовремя, а у службы B происходят огромные колебания — от глубоких опережений графика до жестких опозданий.

2. Какая служба является более предсказуемой?

Служба А является гораздо более предсказуемой, так как у нее минимальное стандартное отклонение (std = 1.41 против 11.40 у B).

3. Какую службу вы выбрали бы для доставки дорогой электроники?

Для дорогой электроники следует выбрать Службу А.

4. Какие статистические показатели повлияли бы на ваше решение?

На решение влияют стандартное отклонение и 90-й процентиль (p90). У службы А 90% клиентов получат дорогой товар в течение 31.6 минут. У службы B показатель p90 составляет 43 минуты — это означает высокий риск того, что дорогостоящий заказ зависнет в пути, заставив клиента долго ждать и волноваться, что критично для клиентского сервиса.

## **Задание 2.4. Корреляция**

In [9]:
np.random.seed(42)

items = np.random.randint(1, 10, 100)
delivery_time_correlated = (
 20
 + 3 * items
 + np.random.normal(0, 3, 100)
)
# 1. Рассчитываем матрицу корреляции Пирсона
corr_matrix = np.corrcoef(items, delivery_time_correlated)

# 2. Извлекаем коэффициент r (значение на пересечении разных признаков)
r_value = corr_matrix[0, 1]

print(f"1. Матрица корреляции:\n{corr_matrix}\n")
print(f"2. Извлеченный коэффициент корреляции Пирсона r: {r_value:.4f}")

1. Матрица корреляции:
[[1.         0.93432937]
 [0.93432937 1.        ]]

2. Извлеченный коэффициент корреляции Пирсона r: 0.9343


**Анализ связи**

**Направление связи:** Связь является положительной (прямой), так как коэффициент (r > 0). Это означает, что с ростом количества товаров в заказе время доставки также имеет тенденцию увеличиваться.

**Сила линейной связи:** По шкале Чеддока значение (r = 0.934) указывает на высокую (очень сильную) линейную взаимосвязь между признаками.

4. Возможное объяснение зависимости с точки зрения логистики: Увеличение количества товаров напрямую затягивает внутренние складские и логистические процессы: сборщику требуется больше времени, чтобы обойти ячейки и найти все позиции; упаковка крупного заказа занимает больше минут; курьеру физически тяжелее нести объемный заказ от машины до двери клиента, что замедляет темп его движения.

**Можно ли утверждать: «Увеличение количества товаров является причиной увеличения времени доставки»?**

Нет, на основании одного лишь высокого коэффициента корреляции утверждать это нельзя. Корреляция фиксирует исключительно математическую согласованность изменений двух величин, но сама по себе никогда не доказывает причинно-следственную связь. Чтобы доказать, что именно количество товаров вызывает задержку, необходим глубокий контролируемый эксперимент с исключением всех побочных факторов.

## **Задание 2.5. Корреляция и скрытые факторы**

1. Почему высокая корреляция не доказывает наличие причинно-следственной связи?

Два показателя могут демонстрировать сильную синхронную связь не потому, что один порождает другой, а потому что они оба одновременно зависят от третьего, внешнего фактора. Это явление называется ложной корреляцией.

2. Фактор, который может влиять одновременно на оба показателя:

Таким скрытым фактором является температура воздуха (наступление жаркого лета).

3. Механизм возникновения наблюдаемой корреляции:

Когда наступает сильная летняя жара, люди начинают массово скупать мороженое, чтобы охладиться (растет первый показатель). Одновременно с этим кратно увеличивается число людей, отправляющихся отдыхать на пляжи и водоемы, что из-за несоблюдения правил безопасности ведет к росту несчастных случаев на воде (растет второй показатель). Математически показатели растут синхронно, порождая корреляцию, хотя покупка пломбира никак физически не вредит пловцам.

4. Собственный пример ложной корреляции из технологий/бизнеса:

В e-commerce может наблюдаться сильная положительная корреляция между количеством отправленных писем техподдержки и общей выручкой интернет-магазина. Скрытый фактор — сезонные распродажи (например, Черная пятница). В этот период резко возрастает поток заказов и денег (выручка растет), но из-за наплыва людей параллельно ломаются интерфейсы или возникают задержки, что генерирует лавину жалоб в саппорт (объем писем растет). Будет ошибкой заявить, что если завалить клиентов письмами техподдержки, то выручка поползет вверх.

# Блок 3. Bootstrap и статистические гипотезы

## **Задание 3.1. Bootstrap и доверительный интервал**

In [10]:
x = np.array([28, 31, 29, 45, 32, 30])

x_mean = np.mean(x)
print(f"Среднее исходной выборки: {x_mean}\n")

B = 5000
bootstrap_mean = np.empty(B)

for i in range(B):
  bootstrap_sample = np.random.choice(x, size=len(x), replace=True)
  bootstrap_mean[i] = np.mean(bootstrap_sample)

p2_5 = np.quantile(bootstrap_mean, 0.025)
p97_5 = np.quantile(bootstrap_mean, 0.975)

print(f"Центральный 95%-й доверительный интервал для среднего:")
print(f"[{p97_5:.2f}; {p2_5:.2f}]")

Среднее исходной выборки: 32.5

Центральный 95%-й доверительный интервал для среднего:
[37.67; 29.17]


1. Почему Bootstrap используется именно с выборкой с возвращением?

Если извлекать элементы без возвращения, мы каждый раз будем получать одну и ту же исходную выборку просто перемешанную. Среднее значение у всех 5000 перестановок будет абсолютно одинаковым. Извлечение с возвращением имитирует случайный выбор из генеральной совокупности, позволяя элементам дублироваться или полностью отсутствовать в подвыборках. Это и создает необходимую вариативность для оценки погрешности.

2. Как наличие наблюдения 45 влияет на Bootstrap-распределение среднего?

Значение 45 — это явный правый выброс для нашей маленькой группы. Поскольку выборка случайная, в некоторых Bootstrap-наборах число 45 может не выпасть ни разу (тогда среднее выборки резко упадет), а в других — выпасть 2, 3 или даже 4 раза подряд (тогда среднее подскочит). Из-за этого выброса Bootstrap-распределение средних значений получается растянутым (скошенным) в правую сторону.

3. Почему результаты Bootstrap могут заметно изменяться при использовании очень маленькой исходной выборки?

Потому что ресэмплинг полностью зависит от состава исходных данных. При выборке из 6 элементов каждый из них составляет весомую долю (16.6%). Добавление или исчезновение всего одного уникального значения кардинально меняет форму итогового распределения. Bootstrap не добавляет новой информации, он лишь «размножает» имеющуюся, поэтому на микровыборках он крайне чувствителен к случайному шуму.

4. Можно ли утверждать: «Вероятность того, что истинное среднее находится внутри рассчитанного 95%-го доверительного интервала, равна 95%»?

Нет, классическая (частотная) статистика запрещает такую формулировку. Истинное среднее генеральной совокупности — это фиксированная, константная величина. Она либо уже лежит внутри конкретного интервала (вероятность 1), либо не лежит там (вероятность 0).Корректная интерпретация: «Если мы проведем этот эксперимент и построим такие доверительные интервалы 100 раз на разных случайных выборках, то в 95 случаях из 100 эти интервалы накроют истинное среднее значение, а в 5 случаях — промахнутся».

## **Задание 3.2. Формулировка гипотез и ошибки I и II рода**

**Ответы на задания и вопросы**

1. Нулевая гипотеза (H_0): Новый алгоритм распределения заказов не уменьшил среднее время доставки. Истинные средние значения в группах равны (\mu_A = \mu_B) или в группе B время даже больше (\mu_A <= \mu_B). Любые замеченные улучшения — это чистая случайность.

2. Альтернативная гипотеза (H_1): Новый алгоритм успешно уменьшил среднее время доставки. Истинное среднее время доставки в группе с новым алгоритмом строго меньше, чем в контрольной (\mu_A > \mu_B).

3. Объяснение ошибок применительно к эксперименту:

- Ошибка I рода: Мы заявляем, что новый алгоритм работает и снижает время доставки (H_0 отвергнута), хотя на самом деле он бесполезен, а зафиксированная разница была случайным шумом.

- Ошибка II рода: Мы заявляем, что новый алгоритм не показал значимых результатов и мы не внедряем его (H_0 не отвергнута), хотя в действительности он был эффективен и реально сокращал время доставки.

4. Сопоставление с терминами:

- Ложная тревога — это Ошибка I рода (мы подняли тревогу, внедрили алгоритм, но эффект ложный).

- Пропущенный эффект — это Ошибка II рода (мы упустили классное решение и не заметили реальный эффект).

**Ситуация для анализа**

Какая ошибка может привести к прямым напрасным расходам компании в 5 млн рублей в год?

К напрасным расходам приведет Ошибка I рода (ложная тревога). Если аналитики допустят эту ошибку, компания поверит в мнимую эффективность алгоритма, полноценно раскатает его на продакшн и начнет ежегодно тратить по 5 миллионов рублей на поддержку системы, которая физически никак не улучшает скорость доставки.

## **Задание 3.3. p-value и принятие решения**

**Ответы на задания и ситуации**

1. Определение p-value: Это вероятность получить такие же или еще более экстремальные различия между группами при условии, что нулевая гипотеза абсолютно верна (эффекта нет).

2. Почему утверждение о «вероятности истинности нулевой гипотезы в 0.1%» неверно?

- Потому что p-value рассчитывается исходя из предположения, что нулевая гипотеза уже на 100% верна. Он оценивает вероятность получить такие данные, а не вероятность самой гипотезы.

3. Является ли результат статистически значимым при α = 0.05?

- Да, является, так как полученный p-value = 0.001 строго меньше установленного уровня значимости α = 0.05. Мы имеем полное право отклонить нулевую гипотезу.

**Анализ ситуации (1 000 000 пользователей, выигрыш 3 секунды, затраты 5 млн рублей)**

1. Является ли результат статистически значимым?
- Да, является. На огромной выборке в 1 миллион пользователей математический критерий заметит даже микроскопическое изменение в 3 секунды и выдаст p-value = 0.001.

2. Достаточно ли одного p-value для принятия решения о внедрении?

- Категорически нет. Статистическая значимость (низкий p-value) — это лишь зеленый свет, подтверждающий, что разница в 3 секунды не случайна. Но она ничего не говорит о целесообразности этой разницы для бизнеса.

3. Почему статистическая значимость не обязательно означает практическую?

- Статистическая значимость зависит от объема данных: на гигантских выборках критерии становятся "сверхчувствительными". Практическая (бизнес) значимость — это величина эффекта. Выигрыш в 3 секунды на фоне среднего времени доставки (например, 30 минут) — это ничтожное изменение (≈ 0.1%), которое клиент никогда физически не заметит.

4. Какие дополнительные показатели необходимо учитывать перед раскаткой?

- Финансовый ROI: Сравнить выгоду от удержания клиентов за счет 3 секунд с затратами в 5 млн рублей (очевидно, что 3 секунды не окупят эти деньги).

- Влияние на другие метрики (Guardrail metrics): Не выросла ли нагрузка на сервера, не упала ли маржинальность, не увеличился ли процент сбоев приложения из-за усложнения логики.

**Итоговый продуктовый вывод**

**Рекомендация:** Отказаться от раскатки нового алгоритма распределения заказов.Несмотря на то, что результаты теста строго статистически значимы (p-value = 0.001 < 0.05), обнаруженный размер эффекта (ускорение всего на 3 секунды) не имеет никакой практической ценности для конечного потребителя. С экономической точки зрения внедрение абсолютно нецелесообразно: микроскопический эффект не способен сгенерировать дополнительную прибыль, которая перекрыла бы ежегодные затраты в размере 5 млн рублей на эксплуатацию данного решения.

# Итоговый вопрос лабораторной работы

Если среднее время уменьшилось, но вырос разброс, p90 не изменился, а p95 вырос (при p-value < 0.05), перед рекомендацией я бы задал продуктовой команде следующие критические вопросы:

- Качество данных: Не обусловлено ли уменьшение среднего значения появлением искусственных аномалий (например, массовым закрытием заказов «в один клик» со временем 0 минут из-за бага в приложении)?

- Размер эффекта: На сколько именно минут снизилось среднее время? Стоит ли это снижение ухудшения остальных метрик?

- Распределение результатов и Процентили: Почему вырос разброс и p95? Это означает, что для подавляющего большинства клиентов ничего не поменялось (p90 на месте), но группа самых «несчастных» клиентов (топ-5% по длительности ожидания, p95) начала ждать свои заказы еще дольше. Наш алгоритм оптимизирует простые заказы за счет ухудшения сервиса в сложных кейсах?

- Выбросы: Не появились ли в группе нового алгоритма жесткие выбросы (например, курьеры начали массово застревать на маршрутах)?

- Бизнес-эффект: Как ухудшение p95 скажется на оттоке клиентов и оценках приложения? Готовы ли мы жертвовать лояльностью долго ждущих клиентов ради красивой, но номинальной цифры снижения общего среднего времени?